base_layer/core/src/validation/block_validators/async_validator.rs

```diff
@@ -0,0 +1,432 @@
+                        let output_hash = input.output_hash();
+                        if output_hashes.iter().all(|hash| *hash != output_hash) {
+                            warn!(
+                                target: LOG_TARGET,
+                                "Validation failed due to input: {} which does not exist yet", input
+                            );
+                            not_found_inputs.push(output_hash);
+                        }
+                    },
+                    Err(err) => return Err(err),
+                    _ => {},
+                }
+
+                // Once we've found unknown inputs, the aggregate data will be discarded and there is no reason to run
+                // the tari script
+                if not_found_inputs.is_empty() {
+                    // lets count up the input script public keys
+                    aggregate_input_key = aggregate_input_key + input.run_and_verify_script(&commitment_factory)?;
+                    commitment_sum = &commitment_sum + &input.commitment;
+                }
+            }
+
+            if !not_found_inputs.is_empty() {
+                return Err(ValidationError::UnknownInputs(not_found_inputs));
+            }
+
+            Ok(InputValidationData {
+                inputs,
+                aggregate_input_key,
+                commitment_sum,
+            })
+        })
+        .into()
+    }
+
+    fn start_output_validation(
+        &self,
+        header: &BlockHeader,
+        outputs: Vec<TransactionOutput>,
+    ) -> AbortOnDropJoinHandle<Result<OutputValidationData, ValidationError>> {
+        let height = header.height;
+        let num_outputs = outputs.len();
+        let concurrency = cmp::min(self.concurrency, num_outputs);
+        let queue = Arc::new(Mutex::new(outputs.into_iter().enumerate().collect::<Vec<_>>()));
+        let bypass_range_proof_verification = self.bypass_range_proof_verification;
+        if bypass_range_proof_verification {
+            info!(target: LOG_TARGET, "Range proof verification will be bypassed!")
+        }
+
+        debug!(
+            target: LOG_TARGET,
+            "Using {} worker(s) to validate #{} ({} output(s))", concurrency, height, num_outputs
+        );
+        let mut output_tasks = (0..concurrency)
+            .map(|_| {
+                let queue = queue.clone();
+                let range_proof_prover = self.factories.range_proof.clone();
+                let db = self.db.inner().clone();
+                task::spawn_blocking(move || {
+                    let db = db.db_read_access()?;
+                    let mut aggregate_sender_offset = PublicKey::default();
+                    let mut commitment_sum = Commitment::default();
+                    let mut outputs = Vec::new();
+                    let mut coinbase_index = None;
+                    while let Some((orig_idx, output)) = queue.lock().expect("lock poisoned").pop() {
```

Comment:
It is not possible for the lock to get poisoned since it only unlocks while pop is running (if you didn't know lock poisoning is when a panic occurs while the lock guard exists i.e. the lock guard isn't safely dropped)

Reply:
Fair enough